### PR TITLE
proto: move keys to oneof 

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

this pr migrates entirely to keys oneof. 

the pubkey interface also added a type function. Before amino encoded the bytes function with the type, now since the bytes method returns []byte representation the type of the key is not known. This is meant to provide a way to check the key type

Closes: #2204

### CODE REVIEW

File: crypto/secp256k1/secp256k1.go [hunk 1]
```diff
@@ -140,6 +144,10 @@ func (pubKey PubKey) String() string {
 	return fmt.Sprintf("PubKeySecp256k1{%X}", []byte(pubKey))
 }
 
+func (pubKey PubKey) Type() string {
+	return "secp256k1"
```
Comment (erikgrinaker) [inline crypto/secp256k1/secp256k1.go:148]: Should we have constants for these? Also, the merged JSON changes now register these keys via `RegisterType()`, so we'll still need type identifiers - not sure how that affects these changes.

File: rpc/client/http/http.go [hunk 2]
```diff
@@ -207,6 +207,7 @@ func (c *baseRPCClient) Status() (*ctypes.ResultStatus, error) {
 	if err != nil {
 		return nil, err
 	}
+
```
Comment (erikgrinaker) [inline rpc/client/http/http.go:210]: What's with all of these spurious whitespace changes?

Comment (marbar3778) [reply rpc/client/http/http.go:210]: I just find it nicer with a space above the final return so I add them here and there. I can undo if needed

Comment (erikgrinaker) [reply rpc/client/http/http.go:210]: Nah, don't really mind them, but when they're unrelated to the other changes they just make the diff noisier and make it harder to do backports.

File: rpc/client/evidence_test.go [hunk 3]
```diff
@@ -135,7 +137,10 @@ func TestBroadcastEvidence_DuplicateVoteEvidence(t *testing.T) {
 		err = abci.ReadMessage(bytes.NewReader(qres.Value), &v)
 		require.NoError(t, err, "Error reading query result, value %v", qres.Value)
 
-		require.EqualValues(t, rawpub, v.PubKey.Data, "Stored PubKey not equal with expected, value %v", string(qres.Value))
+		pk, err := cryptoenc.PubKeyFromProto(v.PubKey)
+		assert.NoError(t, err)
```
Comment (melekes) [inline rpc/client/evidence_test.go:141]:
```suggestion
		require.NoError(t, err)
```

File: state/execution_test.go [hunk 4]
```diff
@@ -355,13 +339,15 @@ func TestEndBlockValidatorUpdates(t *testing.T) {
...
 	pubkey := ed25519.GenPrivKey().PubKey()
+	pk, err := cryptoenc.PubKeyToProto(pubkey)
+	require.NoError(t, err)
 	app.ValidatorUpdates = []abci.ValidatorUpdate{
-		{PubKey: types.TM2PB.PubKey(pubkey), Power: 10},
+		{PubKey: pk, Power: 10},
 	}
 
 	state, _, err = blockExec.ApplyBlock(state, blockID, block)
+	fmt.Println("here")
```
Comment (melekes) [inline state/execution_test.go:349]:
```suggestion
```

File: types/protobuf_test.go [hunk 5]
```diff
@@ -13,22 +13,23 @@ import (
...
+	assert.NoError(t, err)
 }
 
-func testABCIPubKey(t *testing.T, pk crypto.PubKey, typeStr string) {
-	abciPubKey := TM2PB.PubKey(pk)
-	pk2, err := PB2TM.PubKey(abciPubKey)
-	assert.Nil(t, err)
+func testABCIPubKey(t *testing.T, pk crypto.PubKey, typeStr string) error {
+	abciPubKey, err := cryptoenc.PubKeyToProto(pk)
+	assert.NoError(t, err)
```
Comment (melekes) [inline types/protobuf_test.go:28]:
```suggestion
	require.NoError(t, err)
```

**Pull request merged**